File: answers/leetcode/Insertion Sort List/Insertion Sort List.cpp
//@result 21 / 21 test cases passed. Status: Accepted Runtime: 88 ms Submitted: 0 minutes ago You are here! Your runtime beats 20.73% of cpp submissions.

class Solution {
public:
	ListNode *insertionSortList(ListNode *head) {
		if (NULL == head)
		{
			return head;
		}
		ListNode *parent = head, *dull(new ListNode(INT_MIN));
		dull->next = head;
		while (NULL != parent && NULL != parent->next)
		{
			ListNode *target = parent->next;
			ListNode *insert = dull;
			while (NULL != insert && NULL != insert->next && insert->next->val < target->val)
			{
				insert = insert->next;
			}
			if (NULL == insert)
			{
				break;
			}
			if (insert->next != target)
			{
				ListNode *temp = insert->next;
				insert->next = target;
				parent->next = target->next;
				target->next = temp;
			}
			else // if inserted, do not increase iterator
			{
				parent = parent->next;
			}
		}
		return dull->next;
	}
};

File: answers/leetcode/Insertion Sort List/Insertion Sort List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode insertionSortList(ListNode head) {
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode sorted = new ListNode(-1);
        while (dummy.next != null) {
            ListNode current = dummy.next;
            dummy.next = dummy.next.next;
            ListNode target = sorted;
            while (target.next != null && current.val > target.next.val) {
                target = target.next;
            }
            current.next = target.next;
            target.next = current;
        }
        return sorted.next;
    }
}